reference/test/solver/bicg_kernels.cpp

```diff
@@ -0,0 +1,270 @@
+    auto b = gko::initialize<Mtx>({13.0, 7.0, 1.0}, exec);
+
+    auto x = gko::initialize<Mtx>({0.0, 0.0, 0.0}, exec);
+
+    solver->apply(b.get(), x.get());
+
+
+    GKO_ASSERT_MTX_NEAR(x, l({1.0, 3.0, 2.0}), 1e-14);
+}
+
+
+double infNorm(gko::matrix::Dense<> *mat, size_t col = 0)
+{
+    using std::abs;
+    double norm = 0.0;
+    for (size_t i = 0; i < mat->get_size()[0]; ++i) {
+        double absEntry = abs(mat->at(i, col));
+        if (norm < absEntry) norm = absEntry;
+    }
+    return norm;
+}
+
+
+TEST_F(Bicg, SolvesMultipleDenseSystemForDivergenceCheck)
+{
+    auto solver = bicg_factory_big->generate(mtx_big);
+    auto b1 = gko::initialize<Mtx>(
+        {1300083.0, 1018120.5, 906410.0, -42679.5, 846779.5, 1176858.5}, exec);
+    auto b2 = gko::initialize<Mtx>(
+        {886630.5, -172578.0, 684522.0, -65310.5, 455487.5, 607436.0}, exec);
+
+    auto x1 = gko::initialize<Mtx>({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, exec);
+    auto x2 = gko::initialize<Mtx>({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, exec);
+
+    auto bc = Mtx::create(exec, gko::dim<2>{mtx_big->get_size()[0], 2});
+    auto xc = Mtx::create(exec, gko::dim<2>{mtx_big->get_size()[1], 2});
+    for (size_t i = 0; i < bc->get_size()[0]; ++i) {
+        bc->at(i, 0) = b1->at(i);
+        bc->at(i, 1) = b2->at(i);
+
+        xc->at(i, 0) = x1->at(i);
+        xc->at(i, 1) = x2->at(i);
+    }
+
+    solver->apply(b1.get(), x1.get());
+    solver->apply(b2.get(), x2.get());
+    solver->apply(bc.get(), xc.get());
+    auto mergedRes = Mtx::create(exec, gko::dim<2>{b1->get_size()[0], 2});
+    for (size_t i = 0; i < mergedRes->get_size()[0]; ++i) {
+        mergedRes->at(i, 0) = x1->at(i);
+        mergedRes->at(i, 1) = x2->at(i);
+    }
+
+    auto alpha = gko::initialize<Mtx>({1.0}, exec);
+    auto beta = gko::initialize<Mtx>({-1.0}, exec);
+
+    auto residual1 = Mtx::create(exec, b1->get_size());
+    residual1->copy_from(b1.get());
+    auto residual2 = Mtx::create(exec, b2->get_size());
+    residual2->copy_from(b2.get());
+    auto residualC = Mtx::create(exec, bc->get_size());
+    residualC->copy_from(bc.get());
+
+    mtx_big->apply(alpha.get(), x1.get(), beta.get(), residual1.get());
+    mtx_big->apply(alpha.get(), x2.get(), beta.get(), residual2.get());
+    mtx_big->apply(alpha.get(), xc.get(), beta.get(), residualC.get());
+
+    double normS1 = infNorm(residual1.get());
+    double normS2 = infNorm(residual2.get());
+    double normC1 = infNorm(residualC.get(), 0);
+    double normC2 = infNorm(residualC.get(), 1);
+    double normB1 = infNorm(b1.get());
+    double normB2 = infNorm(b2.get());
+
+    // make sure that all combined solutions are as good or better than the
+    // single solutions
+    ASSERT_LE(normC1 / normB1, normS1 / normB1 + 1e-14);
+    ASSERT_LE(normC2 / normB2, normS2 / normB2 + 1e-14);
+
+    // Not sure if this is necessary, the assertions above should cover what
+    // is needed.
+    GKO_ASSERT_MTX_NEAR(xc, mergedRes, 1e-14);
```

Comment:
This is actually a really big test case. I know that the Bicgstab test is also that big, but I would like to make some improvements here so it somewhat follows our AAA rule.
It looks like we could make it Arrange Act Arrange Act Assert out of it, therefore, needing only 4 empty lines in this test.
I would do the first empty line in front of the `apply` calls (meaning all initializations before that should not have any empty lines), the next before the `alpha` initialization (since we allocate stuff here again) and the next in front of the `apply` again. The `infNorm` and the `ASSERT`s can be grouped together in my opinion, so the last empty line should be in front of the first `infNorm`.